appc/appc.go

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package appc implements App Connectors.
+package appc
+
+import (
+	"expvar"
+	"log"
+	"net"
+	"net/netip"
+	"sync"
+	"time"
+
+	"golang.org/x/net/dns/dnsmessage"
+	"tailscale.com/appctype"
+	"tailscale.com/metrics"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/ipproto"
+	"tailscale.com/types/nettype"
+	"tailscale.com/util/clientmetric"
+	"tailscale.com/util/mak"
+)
+
+var tsMBox = dnsmessage.MustNewName("support.tailscale.com.")
+
+// target describes the predicates which route some inbound
+// traffic to the app connector to a specific handler.
+type target struct {
+	Dest     netip.Prefix
+	Matching tailcfg.ProtoPortRange
+}
+
+// Server implements an App Connector.
+type Server struct {
+	m appcMetrics
+
+	mu         sync.RWMutex // mu guards following fields
+	connectors map[appctype.ConfigID]connector
+}
+
+type appcMetrics struct {
+	dnsResponses   expvar.Int
+	dnsFailures    expvar.Int
+	tcpConns       expvar.Int
+	sniConns       expvar.Int
+	unhandledConns expvar.Int
+}
+
+// InstallMetrics publishes metrics about the operation of the app connector.
+// This method must only be called once.
+func (s *Server) InstallMetrics() {
+	stats := new(metrics.Set)
+
+	stats.Set("tls_sessions", &s.m.sniConns)
+	clientmetric.NewCounterFunc("sniproxy_tls_sessions", s.m.sniConns.Value)
+
+	stats.Set("tcp_sessions", &s.m.tcpConns)
+	clientmetric.NewCounterFunc("sniproxy_tcp_sessions", s.m.tcpConns.Value)
+
+	stats.Set("dns_responses", &s.m.dnsResponses)
+	clientmetric.NewCounterFunc("sniproxy_dns_responses", s.m.dnsResponses.Value)
+
+	stats.Set("dns_failed", &s.m.dnsFailures)
+	clientmetric.NewCounterFunc("sniproxy_dns_failed", s.m.dnsFailures.Value)
+
+	expvar.Publish("sniproxy", stats)
+}
+
+// Configure applies the provided configuration to the app connector.
+func (s *Server) Configure(cfg *appctype.AppConnectorConfig) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.connectors = makeConnectorsFromConfig(cfg)
+}
+
+// HandleTCPFlow implements tsnet.FallbackTCPHandler.
+func (s *Server) HandleTCPFlow(src, dst netip.AddrPort) (handler func(net.Conn), intercept bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, c := range s.connectors {
+		if handler, intercept := c.handleTCPFlow(src, dst, &s.m); intercept {
+			return handler, intercept
+		}
+	}
+
+	return nil, false
+}
+
+// HandleDNS handles a DNS request to the app connector.
+func (s *Server) HandleDNS(c nettype.ConnPacketConn) {
```

Comment:
We've been observing trouble with oversized DNS responses in the main product, which has led to us fixing the TCP DNS path. If we move to peer API it's not a concern, but if we stick with DNS we'll need a TCP path as well.

Reply:
:+1: